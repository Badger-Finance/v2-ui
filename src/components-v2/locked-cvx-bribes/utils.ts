import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
