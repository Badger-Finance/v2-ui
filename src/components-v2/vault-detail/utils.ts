import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
