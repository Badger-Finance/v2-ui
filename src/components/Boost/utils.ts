import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
