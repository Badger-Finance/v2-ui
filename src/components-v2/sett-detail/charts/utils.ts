import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
