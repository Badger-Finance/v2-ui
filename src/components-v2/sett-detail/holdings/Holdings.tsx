import React from 'react';
import { Grid, makeStyles, Tooltip, Typography, useMediaQuery, useTheme } from '@material-ui/core';
import { HoldingItem } from './HoldingItem';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../../mobx/store-context';
import { SettBalance } from '../../../mobx/model/setts/sett-balance';
import { Sett } from '../../../mobx/model/setts/sett';
import { HoldingsActionButtons } from './HoldingsActionButtons';
import { NoHoldings } from './NoHoldings';
import { BadgerSett } from '../../../mobx/model/vaults/badger-sett';
import { TokenBalance } from 'mobx/model/tokens/token-balance';
import { hasBalance } from '../utils';
import { TokenDistribution } from './TokenDistribution';
import HelpIcon from '@material-ui/icons/Help';

const useStyles = makeStyles((theme) => ({
	settInfoTitle: {
		fontSize: 24,
		fontWeight: 500,
	},
	helpIcon: {
		fontSize: 16,
		marginLeft: theme.spacing(0.5),
		color: 'rgba(255, 255, 255, 0.3)',
	},
}));

interface Props {
	sett: Sett;
	badgerSett: BadgerSett;
	tokenBalance: TokenBalance;
	settBalance: SettBalance;
}

export const Holdings = observer(
	({ tokenBalance, settBalance, sett, badgerSett }: Props): JSX.Element => {
		const { setts } = React.useContext(StoreContext);
		const isMediumSizeScreen = useMediaQuery(useTheme().breakpoints.up('sm'));
		const classes = useStyles();

		if (!hasBalance(settBalance) && !TokenBalance.hasBalance(tokenBalance)) {
			return (
				<Grid container>
					<NoHoldings sett={sett} badgerSett={badgerSett} />
				</Grid>
			);
		}

		const { earnedBalance, earnedValue, balance, value } = settBalance;
		const logo = `/assets/icons/${sett.asset.toLowerCase()}.png`;

		const depositToken = setts.getToken(sett.underlyingToken);
		const decimals = depositToken?.decimals || 18;

		return (
			<Grid container>
				<Grid container>
					<Typography className={classes.settInfoTitle}>Your Vault Info</Typography>
				</Grid>
				<Grid container spacing={1}>
					<Grid item xs={12} sm>
						<HoldingItem
							name="Your Sett Vaultâ€™s Total"
							logo={logo}
							balance={balance}
							value={value}
							decimals={decimals}
							helpIcon={
								<Tooltip
									enterTouchDelay={0}
									arrow
									placement="top"
									title={<TokenDistribution settBalance={settBalance} />}
								>
									<HelpIcon className={classes.helpIcon} />
								</Tooltip>
							}
						/>
					</Grid>
					<Grid item xs={12} sm>
						<HoldingItem
							name="Total Earned"
							logo={logo}
							balance={earnedBalance}
							value={earnedValue}
							decimals={decimals}
						/>
					</Grid>
					{isMediumSizeScreen && (
						<Grid item xs={12} sm>
							<HoldingsActionButtons />
						</Grid>
					)}
				</Grid>
			</Grid>
		);
	},
);
