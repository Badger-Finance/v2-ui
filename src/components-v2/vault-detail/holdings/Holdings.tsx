import React from 'react';
import { Grid, makeStyles, Typography, useMediaQuery, useTheme } from '@material-ui/core';
import { HoldingItem } from './HoldingItem';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../../mobx/store-context';
import { HoldingsActionButtons } from './HoldingsActionButtons';
import { NoHoldings } from './NoHoldings';
import { TokenDistributionIcon } from './TokenDistributionIcon';
import { VaultDTO, VaultData } from '@badger-dao/sdk';
import { shouldDisplayEarnings } from 'utils/componentHelpers';
import { formatWithoutExtraZeros, numberWithCommas } from '../../../mobx/utils/helpers';
import { useVaultInformation } from 'hooks/useVaultInformation';

const useStyles = makeStyles((theme) => ({
	settInfoTitle: {
		fontSize: 24,
		fontWeight: 500,
	},
	helpIcon: {
		fontSize: 16,
		marginLeft: theme.spacing(0.5),
		color: 'rgba(255, 255, 255, 0.3)',
	},
}));

interface Props {
	vault: VaultDTO;
	userData: VaultData;
}

export const Holdings = observer(({ userData, vault }: Props): JSX.Element | null => {
	const { user, vaults } = React.useContext(StoreContext);
	const isMediumSizeScreen = useMediaQuery(useTheme().breakpoints.up('sm'));
	const classes = useStyles();
	const canDeposit = user.onGuestList(vault);
	const { depositBalance } = useVaultInformation(vault);

	if (depositBalance.tokenBalance.eq(0)) {
		return (
			<Grid container>
				<NoHoldings vault={vault} />
			</Grid>
		);
	}

	const { earnedBalance, earnedValue } = userData;
	const decimals = depositBalance.token.decimals;

	return (
		<Grid container>
			<Grid container>
				<Typography className={classes.settInfoTitle}>Your Vault Info</Typography>
			</Grid>
			<Grid container spacing={1} alignItems="center">
				<Grid item xs={12} sm>
					<HoldingItem
						vault={vault}
						name="Total Deposited"
						balance={depositBalance.balanceDisplay()}
						value={depositBalance.balanceValueDisplay(vaults.vaultsFilters.currency) ?? '0'}
						helpIcon={<TokenDistributionIcon settBalance={userData} />}
					/>
				</Grid>
				{shouldDisplayEarnings(vault, userData) && (
					<Grid item xs={12} sm>
						<HoldingItem
							vault={vault}
							name="Total Earned"
							balance={formatWithoutExtraZeros(earnedBalance, decimals)}
							value={`~$${numberWithCommas(formatWithoutExtraZeros(earnedValue, 2))}`}
						/>
					</Grid>
				)}
				{isMediumSizeScreen && (
					<Grid item xs={12} sm>
						<HoldingsActionButtons canDeposit={canDeposit} />
					</Grid>
				)}
			</Grid>
		</Grid>
	);
});
