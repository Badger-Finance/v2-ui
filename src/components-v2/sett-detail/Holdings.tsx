import React from 'react';
import { Grid } from '@material-ui/core';
import { HoldingItem } from './holdings/HoldingItem';

export const Holdings = (): JSX.Element => {
	return (
		<Grid container spacing={1}>
			<Grid item xs={12} sm>
				<HoldingItem
					name="Your Total"
					logo="/assets/icons/slp-digg-wbtc.png"
					amount="1.21345"
					dollarAmount="~$75,213"
				/>
			</Grid>
			<Grid item xs={12} sm>
				<HoldingItem
					name="Principle"
					logo="/assets/icons/slp-digg-wbtc.png"
					amount="1.01312"
					dollarAmount="~$70,123"
				/>
			</Grid>
			<Grid item xs={12} sm>
				<HoldingItem
					name="Earned"
					logo="/assets/icons/slp-digg-wbtc.png"
					amount="0.21345"
					dollarAmount="~$5,090"
				/>
			</Grid>
		</Grid>
	);
};
