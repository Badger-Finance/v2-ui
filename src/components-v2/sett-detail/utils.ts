import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
