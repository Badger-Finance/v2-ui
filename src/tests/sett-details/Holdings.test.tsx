import React from 'react';
import { checkSnapshot } from 'tests/utils/snapshots';
import { SAMPLE_BADGER_SETT, SAMPLE_SETT, SAMPLE_SETT_BALANCE } from '../utils/samples';
import { Holdings } from '../../components-v2/sett-detail/holdings/Holdings';

describe('Breadcrumb', () => {
	it('displays holdings with no balance', () => {
		checkSnapshot(<Holdings sett={SAMPLE_SETT} badgerSett={SAMPLE_BADGER_SETT} />);
	});

	it('displays holdings with balance', () => {
		checkSnapshot(
			<Holdings sett={SAMPLE_SETT} badgerSett={SAMPLE_BADGER_SETT} settBalance={SAMPLE_SETT_BALANCE} />,
		);
	});
});
