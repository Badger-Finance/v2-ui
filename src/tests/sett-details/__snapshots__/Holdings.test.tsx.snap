// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holdings displays holdings with balance 1`] = `
<div>
  <div
    class="MuiGrid-root-3 MuiGrid-container-4"
  >
    <div
      class="MuiGrid-root-3 MuiGrid-container-4"
    >
      <p
        class="MuiTypography-root-106 makeStyles-settInfoTitle-1 MuiTypography-body1-108"
      >
        Your Vault Info
      </p>
    </div>
    <div
      class="MuiGrid-root-3 MuiGrid-container-4 MuiGrid-spacing-xs-1-26 MuiGrid-align-items-xs-center-12"
    >
      <div
        class="MuiGrid-root-3 MuiGrid-item-5 MuiGrid-grid-xs-12-49 MuiGrid-grid-sm-true-51"
      >
        <div
          class="MuiPaper-root-143 makeStyles-cardContainer-138 MuiPaper-elevation1-147 MuiPaper-rounded-144"
        >
          <div
            class="makeStyles-titleContainer-136"
          >
            <p
              class="MuiTypography-root-106 makeStyles-holdingsName-137 MuiTypography-body1-108"
            >
              Total Deposited
            </p>
            <svg
              aria-hidden="true"
              aria-label="sett token distribution"
              class="MuiSvgIcon-root-183 makeStyles-helpIcon-171"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
              />
            </svg>
          </div>
          <hr
            class="MuiDivider-root-192"
          />
          <div
            class="MuiGrid-root-3 makeStyles-amountsContainer-141 MuiGrid-container-4"
          >
            <div
              class="MuiBox-root-199 MuiBox-root-200 makeStyles-amountText-142"
            >
              <div
                class="makeStyles-logoContainer-139"
              >
                <div
                  class="makeStyles-root-201"
                >
                  <img
                    alt="BBTC logo"
                    class="makeStyles-position-202"
                    height="32"
                    src="/assets/icons/bbtc.svg"
                    width="32"
                  />
                  <img
                    alt="crvRenWSBTC logo"
                    class="makeStyles-position-203"
                    height="32"
                    src="/assets/icons/crvrenwsbtc.svg"
                    width="32"
                  />
                </div>
              </div>
              <div>
                <h5
                  class="MuiTypography-root-106 MuiTypography-h5-115"
                >
                  1.99000
                </h5>
                <p
                  class="MuiTypography-root-106 MuiTypography-body2-107 MuiTypography-colorTextSecondary-132"
                >
                  $26
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Holdings displays holdings with no balance 1`] = `
<div>
  <div
    class="MuiGrid-root-3 MuiGrid-container-4"
  >
    <div
      class="MuiGrid-root-3 MuiGrid-container-4"
    >
      <p
        class="MuiTypography-root-106 makeStyles-settInfoTitle-1 MuiTypography-body1-108"
      >
        Your Vault Info
      </p>
    </div>
    <div
      class="MuiGrid-root-3 MuiGrid-container-4 MuiGrid-spacing-xs-1-26 MuiGrid-align-items-xs-center-12"
    >
      <div
        class="MuiGrid-root-3 MuiGrid-item-5 MuiGrid-grid-xs-12-49 MuiGrid-grid-sm-true-51"
      >
        <div
          class="MuiPaper-root-143 makeStyles-cardContainer-138 MuiPaper-elevation1-147 MuiPaper-rounded-144"
        >
          <div
            class="makeStyles-titleContainer-136"
          >
            <p
              class="MuiTypography-root-106 makeStyles-holdingsName-137 MuiTypography-body1-108"
            >
              Total Deposited
            </p>
            <svg
              aria-hidden="true"
              aria-label="sett token distribution"
              class="MuiSvgIcon-root-183 makeStyles-helpIcon-171"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
              />
            </svg>
          </div>
          <hr
            class="MuiDivider-root-192"
          />
          <div
            class="MuiGrid-root-3 makeStyles-amountsContainer-141 MuiGrid-container-4"
          >
            <div
              class="MuiBox-root-199 MuiBox-root-200 makeStyles-amountText-142"
            >
              <div
                class="makeStyles-logoContainer-139"
              >
                <div
                  class="makeStyles-root-201"
                >
                  <img
                    alt="BBTC logo"
                    class="makeStyles-position-202"
                    height="32"
                    src="/assets/icons/bbtc.svg"
                    width="32"
                  />
                  <img
                    alt="crvRenWSBTC logo"
                    class="makeStyles-position-203"
                    height="32"
                    src="/assets/icons/crvrenwsbtc.svg"
                    width="32"
                  />
                </div>
              </div>
              <div>
                <h5
                  class="MuiTypography-root-106 MuiTypography-h5-115"
                >
                  1.99000
                </h5>
                <p
                  class="MuiTypography-root-106 MuiTypography-body2-107 MuiTypography-colorTextSecondary-132"
                >
                  $26
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
